File: gold-tool/server/textnorm.js
// Port of pipeline/textnorm.py — must stay in lockstep with it (spec §4).
// Normalizes markdown-escape/quote/whitespace/case noise, then checks span
// containment. Source precedence: comment > parent > post (closest wins).

const QUOTE_MAP = {
  "‘": "'", "’": "'", "‚": "'", "‛": "'",
  "“": '"', "”": '"', "„": '"',
  "–": "-", "—": "-", "−": "-",
  " ": " ",
  "…": "...",
};
const QUOTE_RE = new RegExp(`[${Object.keys(QUOTE_MAP).join("")}]`, "g");

const HTML_ENTITIES = { amp: "&", lt: "<", gt: ">", quot: '"', "#39": "'", apos: "'", nbsp: " " };

function unescapeHtml(text) {
  return text.replace(/&(amp|lt|gt|quot|#39|apos|nbsp);/g, (_, e) => HTML_ENTITIES[e]);
}

export function normalize(text) {
  if (!text) return "";
  let t = unescapeHtml(text);
  t = t.replace(/\\/g, "");
  t = t.replace(QUOTE_RE, (ch) => QUOTE_MAP[ch]);
  t = t.replace(/\s+/g, " ").trim();
  return t.toLowerCase();
}

export function spanIn(span, text) {
  if (!span || !text) return false;
  return normalize(text).includes(normalize(span));
}

const SOURCE_PRECEDENCE = ["comment", "parent", "post"];

export function findSpanSource(span, sources) {
  for (const name of SOURCE_PRECEDENCE) {
    if (spanIn(span, sources[name])) return name;
  }
  return null;
}

File: pipeline/textnorm.py
"""Span normalization and source derivation (spec §4).

Spans must be verbatim, but "verbatim" has to survive the noise Reddit and LLMs
introduce: markdown escapes (\\_ \\* \\[), smart quotes, HTML entities, casing and
whitespace differences. `normalize()` folds all of that away; containment is then
checked on normalized text.

`find_span_source()` derives where a span came from, with the spec's precedence
comment > parent > post — the closest text to the author wins.
"""
from __future__ import annotations

import html
import re

_WS_RE = re.compile(r"\s+")

# Folds curly quotes AND straight " onto straight ' — models reliably copy a
# "double-quoted" source phrase back as 'single-quoted' (2026-07-04 audit),
# a punctuation-only verbatim mismatch, not a content change.
_QUOTE_MAP = str.maketrans({
    "‘": "'", "’": "'", "‚": "'", "‛": "'",
    "“": "'", "”": "'", "„": "'",
    '"': "'",
    "–": "-", "—": "-", "−": "-",
    " ": " ",
    "…": "...",
})


def normalize(text: str | None) -> str:
    """Fold escaping/quote/whitespace/case noise; keep word content intact."""
    if not text:
        return ""
    text = html.unescape(text)
    # Backslashes only ever appear as markdown escapes (possibly stacked, e.g. the
    # shrug emoticon ¯\\\_(ツ)\_/¯); they are never semantic for span matching.
    text = text.replace("\\", "")
    text = text.translate(_QUOTE_MAP)
    text = _WS_RE.sub(" ", text).strip()
    return text.casefold()


def span_in(span: str | None, text: str | None) -> bool:
    """True if `span` occurs verbatim (normalized) inside `text`."""
    if not span or not text:
        return False
    return normalize(span) in normalize(text)


# Spec §4 precedence: the closest text to the author wins.
SOURCE_PRECEDENCE = ("comment", "parent", "post")


def find_span_source(span: str | None, sources: dict[str, str | None]) -> str | None:
    """Derive which source text a span was copied from, or None if none match.

    `sources` maps source name -> text, e.g. {"comment": body, "parent": parent_body,
    "post": title + selftext}. Missing/None texts are skipped.
    """
    for name in SOURCE_PRECEDENCE:
        if span_in(span, sources.get(name)):
            return name
    return None
